### ISSUE TITLE
`rails server` doesn't listen on `localhost` on Linux

### ISSUE THREAD

**Issue by schuetzm** [rails/rails#29516]
In the past, after running `rails server` without the `-b` option, it was possible to access the Rails app on http://localhost:3000. Since Rails 5 (or 5.1?), it will bind to the system's hostname instead.

I've tracked this down to the use of the `$HOST` env variable in https://github.com/rails/rails/blob/master/railties/lib/rails/commands/server/server_command.rb, which is used together with `$PORT` to determine the listening options for the webserver. It seems this was done to make it configurable by setting these variables explicitly; however on some Linux variants (at least SuSE), `$HOST` is set by default and is equal to `$HOSTNAME`. This can either cause the server to bind to that hostname, or even, if the hostname doesn't resolve, to refuse to start.

**Comment by mathieujobin** [comment]
using up to date archlinux system.
with fresh rails 5.1.1 app, tested with both webrick and puma.
i don't have this problem
Env `$HOST is not set on my system though

**Comment by y-yagi** [comment]
The rails server started using `HOST` environment variable from 5.1. This has been added in #25688 .

cc: @guilleiguaran  
(Because you were commenting on https://github.com/rails/rails/pull/25688#issuecomment-231434908 )

**Comment by kirs** [comment]
Wouldn't be listening `localhost` more safe than listening current hostname (which could be reachable from the outside)?

**Comment by alexcameron89** [comment]
@kirs Yeah, I think it's supposed to be listening to `localhost`. It looks like this changed with 7e2a79fab66ed92ace12a8bb965007f6bf9e93bb, cc @schneems.

**Comment by schneems** [comment]
Is this fixed on master?

**Comment by kirs** [comment]
It's not. There's more discussion in https://github.com/rails/rails/pull/25688#issuecomment-312816670, but I don't see a way to fix it other than renaming `$HOST` to something else that wouldn't conflict with those Linux variants that are using this variable.

**Comment by schuetzm** [comment]
Tested on `master`, still doesn't work (of course).

**Comment by schuetzm** [comment]
Bug is still there on ```master```.

**Comment by lpar** [comment]
Just hit this bug on 5.1.4.

```
> rails s -p 3000
=> Booting Puma
=> Rails 5.1.4 application starting in production 
=> Run `rails server -h` for more startup options
Puma starting in single mode...
* Version 3.11.0 (ruby 2.4.2-p198), codename: Love Song
* Min threads: 5, max threads: 30
* Environment: production
* Listening on tcp://w541:3000
```

**Comment by y-yagi** [comment]
I have deprecated using `HOST` environment to specify server IP. This issue is correctly fixed in Rails 6.1. 
As that step, Rails 6.0 will have unintended deprecate messages. In that case, please specify `BINDING` environment. Thanks.
